text/0000-xor-urls/0000-xor-urls.md

```diff
@@ -0,0 +1,119 @@
+
+#### MutableData versions
+MutableData’s are versioned and therefore this shall be also accountable in the XOR-URL format to allow any MD URL to (optionally) reference a specific version. This will be also necessary in the future when append-only type of data is made available on the network.
+
+The version value can be used to enforce a specific version to be retrieved, and otherwise fail if that specific version is not found. On the other hand, the latest version will be retrieved if the version value is omitted from the XOR-URL.
+
+#### Paths
+Given that a referenced MutableData can effectively be the root NFS/Files container of the files of a safesite, any MD XOR-URL can also specify a path which needs to be resolved, just like when using the DNS/public-names system to resolve the path in a URL.
+
+#### Browsable content
+Currently, when a public-name URL is resolved to an NFS/Files container which doesn’t have an `index.html` file, the browser simply shows an error stating that the safesite content was not found.
+
+For a MD XOR-URL that doesn’t have a path, and there is no `index.html` file or a `default` file defined (for more details about this please see refer to [this other RFC](https://github.com/joshuef/rfcs/blob/PnsAndResolveableMap/text/0000-RDF-for-public-name-resolution/0000-RDF-for-public-name-resolution.md)), the resolver can return the raw content of the MD (i.e. its key-value entries), and the browser can automatically generate an HTML page which makes the content browsable, generating links to other data when an entry’s key/value is a `safe://` string, in an analogous/similar way to how web servers on the current internet allow browsing on folders/directories.
+
+### XOR-URLs specification
+The following are the main requirements for the encoding to be used to generate the XOR-URLs:
+
+- Be able to support new and different types of base encodings and hash functions for the XOR addresses in the future.
+- Include the content type within the XOR URL which would allow the client app to correctly render the data to the user especially when referencing an ImmutableData.
+
+It is proposed the use of [multiformats](https://github.com/multiformats) and [CID](https://github.com/ipld/cid), which allow us to cover the above requirements. A CID identifier can be used in the XOR-URL for specifying the XOR address part, and have additional parts to support MutableData’s type tag, version, as well as the path and query parts. SAFE URLs can then be defined in the following way (BNF 1-like):
+```
+<safe-url>            := 'safe://' ( <xor-uri> | <public-name-uri> )
+
+<public-name-uri>     :=  [<service> '.'] <public-name> <path-query-fragment>
+
+<xor-uri>             := <immutable-data-uri> | <mutable-data-uri>
+
+<immutable-data-uri>  := <cid> <query-fragment>
+
+<mutable-data-uri>    := <cid> ':' <type-tag> ['+' <content-version>] <path-query-fragment>
```

Comment:
While generating the digest of the multihash, we are using only the XOR name. And then we are adding tagType and version to the URL. For example, `safe://xor-address-of-md:tagType+version/sample.png`. This convention would make it difficult to parse IMO. 
Is it possible to have a json like representation,
```
{
  name: xor-name,//multibase encoded
  tageType: 15000,
  version: ''//string or number
}
```
This could be used to generate the CID like,
```
const getData = (address, tagType) => {
    return new Buffer(JSON.stringify({address, tagType}));
};
const encodedHash = mHash.encode(getData('somedata', 1000), 'sha1');
const cid = new CID(1, 'raw', encodedHash);
const encodedCID = cid.toBaseEncodedString('base32');
```
When decoded we have a defined structure to resolve.

Reply:
It could be, although I thought having the type tag and versions exposed in the URL could be better for the user. The type tag matches quite well if you think of an analogy with ports in the clearnet, and having the version available to the user to change and get a different version of the same content is also I thought could be good, and as mentioned in the RFC if you are willing to get the latest then you omit the version from the URL. Basically if you see the version in the URL explicitly could probably be better for the user as he/she knows which version is being fetched and even change it.